Core/include/Acts/EventData/Measurement.hpp

```diff
@@ -9,387 +9,244 @@
 
-#include <memory>
-#include <ostream>
-#include <type_traits>
-#include <utility>
+#include <cstddef>
+#include <iosfwd>
+#include <variant>
 
 namespace Acts {
 
-// forward declarations
-class Surface;
-
-namespace detail {
-/// @brief Deduction of the measuring geometry object based on the used indices
-template <typename T>
-struct ReferenceObject {};
-template <>
-struct ReferenceObject<BoundIndices> {
-  using type = Surface;
-};
-template <>
-struct ReferenceObject<FreeIndices> {
-  using type = Volume;
-};
-}  // namespace detail
-
-/// @brief base class for Measurements
+/// A measurement of a fixed-size subspace of the full parameters.
 ///
-/// This class describes the measurement of track parameters at a certain
-/// Surface/Volume in the TrackingGeometry.
+/// @tparam source_link_t Source link type to connect to the detector readout
+/// @tparam indices_t Parameter index type, determines the full parameter space
+/// @tparam kSize Size of the parameter subset.
 ///
-/// If the measurement is in local parameters and will not provide localToGlobal
-/// information. It is thus free from any Context.
+/// The measurement is templated on the measurement subspace size and its
+/// content (indices) is defined at runtime. In previous implementations,
+/// the measurement was templated directly on the subspace indices. This lead to
+/// a combinatorial explosion and severe penalty during compilation since all
+/// possible combinations need to be considered for the measurement variant.
+/// However, for the computations only the size of the measurement subspace is
+/// relevant.
 ///
-/// @tparam source_link_t the templated class that allows to link back to
-/// the source used to create this measurement, this can simply be an identifier
-/// or an extended object that allows to navigate back to the original source.
-/// The source link is necessary when e.g. calibrating the object or in other
-/// circumstances where the pure mathematical view of a measurement is not
-/// sufficient.
-/// The name _link_ indicates that there is a relationship to a unique source
-/// object, i.e. if a measurement appears on several tracks, then there must
-/// be a uniquely identifyable common source for those.
+/// The measurement intenionally does not store a pointer/reference to the
+/// explicit reference object in the geometry hierarchy, i.e. the surface or
+/// volume. The reference object can already be identified via the geometry
+/// identifier provided by the source link. Since a measurement
+/// **must** be anchored within the geometry hierarchy, all measurement surfaces
+/// and volumes **must** provide valid geometry identifiers. In all use-cases,
+/// e.g. Kalman filtering, a pointer/reference to the reference object is
+/// available before the measurement is accessed; e.g. the propagator provides
+/// the surface pointer during navigation, which is then used to lookup possible
+/// measurements.
```

Comment:
Here's one concern I am having: what if someone wants to express a bunch of measurements (e.g. a muon track) on a chosen surface that is NOT part of the hierarchy? 
The Kalmanfilter will allow this with the `external measurements` by what do we do here? do we need an extra container?

Reply:
I think that needs a bit more thought anyways. So far, all measurements **must** be anchored in the geometry hierarchy. Both the KF and the CKF use the navigator to provide e.g. the surface and then use the geometry identifier to lookup possible measurements (in the CKF case). This was already a requirement before this PR and removing reference object from the measurement only removes duplicate information (and thus greatly simplifies handling both surface- and volume-based without relying on complicated template hacks).

If you want to support external measurements in the KF that are not anchored in the geometry hierarchy, you would probably need to add support to the navigator anyways. In which case you could provide the external measurements as pairs of reference object and measurement.

Reply:
> 
> If you want to support external measurements in the KF that are not anchored in the geometry hierarchy, you would probably need to add support to the navigator anyways. In which case you could provide the external measurements as pairs of reference object and measurement.

This is what @Corentin-Allaire  is working on, actually.